src/tools/push_av_server/server.py

```diff
@@ -0,0 +1,582 @@
+    for the exact content.
+    """
+    cert_details = req.scope["transport"].get_extra_info("ssl_object").getpeercert()
+
+    logging.debug(f"segment_upload. stream_id={stream_id} file_path:{file_path}")
+
+    if not wd.path("streams", str(stream_id)).exists():
+        raise HTTPException(404, detail="Stream doesn't exists")
+
+    dst = wd.mkdir("streams", str(stream_id), file_path, is_file=True)
+
+    with open(dst.with_suffix(dst.suffix + ".crt"), "w") as f:
+        f.write(json.dumps(cert_details))
+
+    with open(dst, "wb") as f:
+        async for chunk in req.stream():
+            f.write(chunk)
+
+    return Response(status_code=202)
+
+
+@app.get("/streams/probe/{stream_id}/{file_path:path}")
+def ffprobe_check(file_path: str, stream_id: int):
+
+    p = wd.path("streams", str(stream_id), file_path)
+
+    if not p.exists():
+        return HTTPException(404, detail="Stream doesn't exists")
+
+    proc = subprocess.run(
+        ["ffprobe", "-show_streams", "-show_format", "-output_format", "json", str(p.absolute())],
+        capture_output=True
+    )
+
+    if proc.returncode != 0:
+        # TODO Add more details (maybe stderr) to the response
+        return HTTPException(500)
+
+    return json.loads(proc.stdout)
+
+
+# TODO app.post("streams/convert/{stream_id}")
+# Will execute a ffmpeg conversion of the uploaded stream to a regular mp4
+# for easier consumption. Need to double check first if Firefox/Chrome can
+# read DASH media with CMAF tracks. If they can no need to actually convert
+# to a new format.
+
+
+@app.get("/certs", status_code=200)
+def list_certs():
+    server = [f.name for f in pathlib.Path(wd.path("server")).iterdir()]
+    device = [f.name for f in pathlib.Path(wd.path("device")).iterdir()]
+
+    return {"server": server, "device": device}
+
+
+@app.post("/certs/{name}/keypair")
+def create_client_keypair(name: str, override: bool = True):
+    (key, cert, created) = device_hierarchy.gen_keypair(name, override)
+
+    return {key, cert, created}
+
+
+class SignClientCertificate(BaseModel):
+    csr: str
+
+
+@app.post("/certs/{name}/issue")
+def sign_client_certificate(
+    name: str, req: SignClientCertificate, override: bool = True
+):
+    (key, cert, created) = device_hierarchy.gen_cert(name, req.csr, override)
+
+    return {key, cert, created}
+
+
+def run(host: Optional[str], port: Optional[int], working_directory: Optional[str], dns: Optional[str]):
+    # For the Test Harness implementation, will need to find a way to return a reference
+    # to the server running in the background. This function currently doesn't return.
+    with WorkingDirectory(working_directory) as directory:
+
+        # Sharing state with the various endpoints
+        wd = directory
+
+        dns = "localhost" if dns is None else f"{dns}._http._tcp_.local."
+
+        # Create CA hierarchies (for webserver and devices)
+        device_hierarchy = CAHierarchy(directory.mkdir("certs", "device"), "device")
```

Comment:
Since this function modifies global variables, we'll need a `global` keyword to avoid unintentional local allocation. Note that the reads / accesses are fine without the keyword (python 😒 )
 
```suggestion
        global wd, device_hierarchy

        # Sharing state with the various endpoints
        wd = directory

        dns = "localhost" if dns is None else f"{dns}._http._tcp_.local."

        # Create CA hierarchies (for webserver and devices)
        device_hierarchy = CAHierarchy(directory.mkdir("certs", "device"), "device")
```